apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts

```diff
@@ -139,22 +153,47 @@ export class PnpmLinkManager extends BaseLinkManager {
     // appropriate. This folder is usually something like:
     // C:\{uri-encoed-path-to-tgz}\node_modules\{package-name}
 
+    // e.g.:
+    //   file:projects/bentleyjs-core.tgz
+    //   file:projects/build-tools.tgz_dc21d88642e18a947127a751e00b020a
+    //   file:projects/imodel-from-geojson.tgz_request@2.88.0
+    const topLevelDependencyVersion: string | undefined =
+      pnpmShrinkwrapFile!.getTopLevelDependencyVersion(project.tempProjectName);
+
+    if (!topLevelDependencyVersion) {
+      throw new InternalError(`Cannot find top level dependency for "${project.tempProjectName}"` +
+        ` in shrinkwrap.`);
+    }
+
+    const packageId: string | undefined = pnpmShrinkwrapFile!.getPackageId(topLevelDependencyVersion);
+
+    // e.g.: projects\api-documenter.tgz
+    let relativePathToTgzFile: string;
+
+    if (packageId) {
+      relativePathToTgzFile = packageId.slice('file:'.length);
+    } else {
+      relativePathToTgzFile = topLevelDependencyVersion.slice('file:'.length);
+    }
+
     // e.g.: C:\wbt\common\temp\projects\api-documenter.tgz
-    const pathToTgzFile: string = path.join(
-      this._rushConfiguration.commonTempFolder,
-      'projects',
-      `${unscopedTempProjectName}.tgz`);
+    const absolutePathToTgzFile: string = path.join(this._rushConfiguration.commonTempFolder, relativePathToTgzFile);
 
-    // e.g.: C%3A%2Fwbt%2Fcommon%2Ftemp%2Fprojects%2Fapi-documenter.tgz
-    const escapedPathToTgzFile: string = uriEncode(Text.replaceAll(pathToTgzFile, path.sep, '/'));
+    // e.g.:
+    //   C%3A%2Fwbt%2Fcommon%2Ftemp%2Fprojects%2Fapi-documenter.tgz
+    //   C%3A%2Fdev%2Fimodeljs%2Fimodeljs%2Fcommon%2Ftemp%2Fprojects%2Fpresentation-integration-tests.tgz_jsdom@11.12.0
+    //   C%3A%2Fdev%2Fimodeljs%2Fimodeljs%2Fcommon%2Ftemp%2Fprojects%2Fbuild-tools.tgz_2a665c89609864b4e75bc5365d7f8f56
+    const dirNameInLocal: string = uriEncode(Text.replaceAll(absolutePathToTgzFile, path.sep, '/')) +
+      (packageId && packageId.length < topLevelDependencyVersion.length ?
```

Comment:
This is somewhat hard to read. Can you split the `packageId && ...` stuff into an intermediate variable?

Reply:
Will do